File: src/books/reading-progress-service.ts
import type { Database } from "../../database.ts";
import * as bookData from "./data.ts";

export class ReadingProgressService {
    constructor(private database: Database) {}

    /**
     * Update reading progress for a user
     */
    async updateProgress(
        user_id: number, 
        book_id: number, 
        current_page: number, 
        total_pages?: number
    ): Promise<bookData.ReadingProgress> {
        // Validate page numbers
        if (current_page < 0) {
            throw new Error("Current page cannot be negative");
        }
        
        if (total_pages && total_pages <= 0) {
            throw new Error("Total pages must be positive");
        }
        
        if (total_pages && current_page > total_pages) {
            throw new Error("Current page cannot exceed total pages");
        }

        return await bookData.upsertReadingProgress(
            this.database, 
            user_id, 
            book_id, 
            current_page, 
            total_pages
        );
    }

    /**
     * Get reading progress for a specific book
     */
    async getProgress(user_id: number, book_id: number): Promise<bookData.ReadingProgress | null> {
        return await bookData.getReadingProgress(this.database, user_id, book_id);
    }

    /**
     * Get all reading progress for a user
     */
    async getAllProgress(user_id: number): Promise<bookData.ReadingProgress[]> {
        return await bookData.getUserReadingProgress(this.database, user_id);
    }

    /**
     * Get recently read books for a user
     */
    async getRecentlyRead(user_id: number, limit: number = 10): Promise<(bookData.ReadingProgress & bookData.BookWithId)[]> {
        return await bookData.getRecentlyReadBooks(this.database, user_id, limit);
    }

    /**
     * Get reading statistics for a user
     */
    async getReadingStats(user_id: number) {
        return await bookData.getUserReadingStats(this.database, user_id);
    }

    /**
     * Mark a book as completed
     */
    async markAsCompleted(user_id: number, book_id: number): Promise<bookData.ReadingProgress> {
        return await bookData.markBookAsCompleted(this.database, user_id, book_id);
    }

    /**
     * Start reading a book (set to page 1 if no progress exists)
     */
    async startReading(user_id: number, book_id: number, total_pages?: number): Promise<bookData.ReadingProgress> {
        const existingProgress = await this.getProgress(user_id, book_id);
        
        if (existingProgress) {
            // Book already started, return existing progress
            return existingProgress;
        }
        
        // Start from page 1
        return await this.updateProgress(user_id, book_id, 1, total_pages);
    }

    /**
     * Reset reading progress (restart book)
     */
    async resetProgress(user_id: number, book_id: number): Promise<void> {
        await bookData.deleteReadingProgress(this.database, user_id, book_id);
    }

    /**
     * Update progress by percentage (0-100)
     */
    async updateProgressByPercentage(
        user_id: number, 
        book_id: number, 
        percentage: number, 
        total_pages?: number
    ): Promise<bookData.ReadingProgress> {
        if (percentage < 0 || percentage > 100) {
            throw new Error("Percentage must be between 0 and 100");
        }
        
        // If we don't have total pages, get it from existing progress or default
        let actualTotalPages = total_pages;
        if (!actualTotalPages) {
            const existingProgress = await this.getProgress(user_id, book_id);
            actualTotalPages = existingProgress?.total_pages || 100; // Default to 100 pages
        }
        
        const current_page = Math.round((percentage / 100) * actualTotalPages);
        return await this.updateProgress(user_id, book_id, current_page, actualTotalPages);
    }

    /**
     * Get currently reading books (in progress, not completed)
     */
    async getCurrentlyReading(user_id: number): Promise<(bookData.ReadingProgress & bookData.BookWithId)[]> {
        const allProgress = await this.getRecentlyRead(user_id, 50); // Get more to filter
        return allProgress.filter(progress => 
            progress.percentage_complete > 0 && progress.percentage_complete < 100
        );
    }

    /**
     * Get completed books
     */
    async getCompletedBooks(user_id: number): Promise<(bookData.ReadingProgress & bookData.BookWithId)[]> {
        const allProgress = await this.getRecentlyRead(user_id, 100); // Get more to filter
        return allProgress.filter(progress => progress.percentage_complete >= 100);
    }

    /**
     * Get reading streak (days in a row with reading activity)
     */
    async getReadingStreak(user_id: number): Promise<number> {
        const result = await this.database.execute({
            sql: `
                WITH daily_reading AS (
                    SELECT DISTINCT DATE(last_read_at) as read_date
                    FROM reading_progress 
                    WHERE user_id = $user_id
                    ORDER BY read_date DESC
                ),
                consecutive_days AS (
                    SELECT 
                        read_date,
                        ROW_NUMBER() OVER (ORDER BY read_date DESC) as row_num,
                        DATE(read_date, '+' || (ROW_NUMBER() OVER (ORDER BY read_date DESC) - 1) || ' days') as expected_date
                    FROM daily_reading
                )
                SELECT COUNT(*) as streak
                FROM consecutive_days
                WHERE read_date = expected_date
                AND read_date >= DATE('now', '-' || (
                    SELECT COUNT(*) FROM consecutive_days
                ) || ' days')
            `,
            args: { user_id }
        });

        return result.rows[0]?.streak as number || 0;
    }

    /**
     * Get reading activity for the last 30 days
     */
    async getReadingActivity(user_id: number, days: number = 30): Promise<{ date: string; pages_read: number; books_read: number }[]> {
        const result = await this.database.execute({
            sql: `
                SELECT 
                    DATE(last_read_at) as date,
                    SUM(current_page) as pages_read,
                    COUNT(DISTINCT book_id) as books_read
                FROM reading_progress 
                WHERE user_id = $user_id 
                AND last_read_at >= DATE('now', '-$days days')
                GROUP BY DATE(last_read_at)
                ORDER BY date DESC
            `,
            args: { user_id, days }
        });

        return result.rows.map(row => ({
            date: row.date as string,
            pages_read: row.pages_read as number,
            books_read: row.books_read as number
        }));
    }
}

File: src/books/reading_progress_service.zig
const std = @import("std");
const sqlite = @import("sqlite");

const ReadingProgress = @import("../books.zig").ReadingProgress;

pub const ReadingProgressService = struct {
    db: *sqlite.Db,
    allocator: std.mem.Allocator,

    pub fn init(db: *sqlite.Db, allocator: std.mem.Allocator) ReadingProgressService {
        return .{
            .db = db,
            .allocator = allocator,
        };
    }

    pub fn getProgress(self: *ReadingProgressService, user_id: i64, book_id: i64) !?ReadingProgress {
        const query =
            \\SELECT id, user_id, book_id, current_page, total_pages, percentage_complete, last_read_at, completed_at, created_at, updated_at
            \\FROM reading_progress
            \\WHERE user_id = ? AND book_id = ?
        ;

        var stmt = try self.db.prepare(query);
        defer stmt.deinit();

        try stmt.bind(.{ user_id, book_id });

        if (try stmt.step()) {
            return ReadingProgress{
                .id = stmt.columnInt64(0),
                .user_id = stmt.columnInt64(1),
                .book_id = stmt.columnInt64(2),
                .current_page = stmt.columnInt64(3),
                .total_pages = if (stmt.columnType(4) == .Null) null else stmt.columnInt64(4),
                .percentage_complete = stmt.columnDouble(5),
                .last_read_at = stmt.columnInt64(6),
                .completed_at = if (stmt.columnType(7) == .Null) null else stmt.columnInt64(7),
                .created_at = stmt.columnInt64(8),
                .updated_at = stmt.columnInt64(9),
            };
        }

        return null;
    }

    pub fn updateProgress(self: *ReadingProgressService, user_id: i64, book_id: i64, current_page: i64, total_pages: ?i64) !ReadingProgress {
        const now = std.time.timestamp();
        
        const percentage: f64 = if (total_pages) |total| 
            @as(f64, @floatFromInt(current_page)) / @as(f64, @floatFromInt(total)) * 100.0 
        else 
            0.0;

        // Upsert the progress
        const query =
            \\INSERT INTO reading_progress (user_id, book_id, current_page, total_pages, percentage_complete, last_read_at, updated_at)
            \\VALUES (?, ?, ?, ?, ?, ?, ?)
            \\ON CONFLICT(user_id, book_id) DO UPDATE SET
            \\  current_page = excluded.current_page,
            \\  total_pages = excluded.total_pages,
            \\  percentage_complete = excluded.percentage_complete,
            \\  last_read_at = excluded.last_read_at,
            \\  updated_at = excluded.updated_at
        ;

        var stmt = try self.db.prepare(query);
        defer stmt.deinit();

        try stmt.bind(.{ user_id, book_id, current_page, total_pages, percentage, now, now });
        try stmt.exec();

        return (try self.getProgress(user_id, book_id)).?;
    }

    pub fn startReading(self: *ReadingProgressService, user_id: i64, book_id: i64, total_pages: ?i64) !ReadingProgress {
        return self.updateProgress(user_id, book_id, 0, total_pages);
    }

    pub fn markAsCompleted(self: *ReadingProgressService, user_id: i64, book_id: i64) !ReadingProgress {
        const now = std.time.timestamp();

        const query =
            \\UPDATE reading_progress 
            \\SET percentage_complete = 100.0, completed_at = ?, last_read_at = ?, updated_at = ?
            \\WHERE user_id = ? AND book_id = ?
        ;

        var stmt = try self.db.prepare(query);
        defer stmt.deinit();

        try stmt.bind(.{ now, now, now, user_id, book_id });
        try stmt.exec();

        return (try self.getProgress(user_id, book_id)).?;
    }

    pub fn resetProgress(self: *ReadingProgressService, user_id: i64, book_id: i64) !void {
        const query =
            \\DELETE FROM reading_progress WHERE user_id = ? AND book_id = ?
        ;

        var stmt = try self.db.prepare(query);
        defer stmt.deinit();

        try stmt.bind(.{ user_id, book_id });
        try stmt.exec();
    }

    pub fn getAllProgress(self: *ReadingProgressService, user_id: i64) !std.ArrayList(ReadingProgress) {
        var progress_list = std.ArrayList(ReadingProgress).init(self.allocator);
        errdefer progress_list.deinit();

        const query =
            \\SELECT id, user_id, book_id, current_page, total_pages, percentage_complete, last_read_at, completed_at, created_at, updated_at
            \\FROM reading_progress
            \\WHERE user_id = ?
            \\ORDER BY last_read_at DESC
        ;

        var stmt = try self.db.prepare(query);
        defer stmt.deinit();

        try stmt.bind(.{user_id});

        while (try stmt.step()) {
            try progress_list.append(ReadingProgress{
                .id = stmt.columnInt64(0),
                .user_id = stmt.columnInt64(1),
                .book_id = stmt.columnInt64(2),
                .current_page = stmt.columnInt64(3),
                .total_pages = if (stmt.columnType(4) == .Null) null else stmt.columnInt64(4),
                .percentage_complete = stmt.columnDouble(5),
                .last_read_at = stmt.columnInt64(6),
                .completed_at = if (stmt.columnType(7) == .Null) null else stmt.columnInt64(7),
                .created_at = stmt.columnInt64(8),
                .updated_at = stmt.columnInt64(9),
            });
        }

        return progress_list;
    }

    pub fn getRecentlyRead(self: *ReadingProgressService, user_id: i64, limit: i64) !std.ArrayList(ReadingProgress) {
        var progress_list = std.ArrayList(ReadingProgress).init(self.allocator);
        errdefer progress_list.deinit();

        const query =
            \\SELECT id, user_id, book_id, current_page, total_pages, percentage_complete, last_read_at, completed_at, created_at, updated_at
            \\FROM reading_progress
            \\WHERE user_id = ?
            \\ORDER BY last_read_at DESC
            \\LIMIT ?
        ;

        var stmt = try self.db.prepare(query);
        defer stmt.deinit();

        try stmt.bind(.{ user_id, limit });

        while (try stmt.step()) {
            try progress_list.append(ReadingProgress{
                .id = stmt.columnInt64(0),
                .user_id = stmt.columnInt64(1),
                .book_id = stmt.columnInt64(2),
                .current_page = stmt.columnInt64(3),
                .total_pages = if (stmt.columnType(4) == .Null) null else stmt.columnInt64(4),
                .percentage_complete = stmt.columnDouble(5),
                .last_read_at = stmt.columnInt64(6),
                .completed_at = if (stmt.columnType(7) == .Null) null else stmt.columnInt64(7),
                .created_at = stmt.columnInt64(8),
                .updated_at = stmt.columnInt64(9),
            });
        }

        return progress_list;
    }

    pub fn getCurrentlyReading(self: *ReadingProgressService, user_id: i64) !std.ArrayList(ReadingProgress) {
        var progress_list = std.ArrayList(ReadingProgress).init(self.allocator);
        errdefer progress_list.deinit();

        const query =
            \\SELECT id, user_id, book_id, current_page, total_pages, percentage_complete, last_read_at, completed_at, created_at, updated_at
            \\FROM reading_progress
            \\WHERE user_id = ? AND completed_at IS NULL AND current_page > 0
            \\ORDER BY last_read_at DESC
        ;

        var stmt = try self.db.prepare(query);
        defer stmt.deinit();

        try stmt.bind(.{user_id});

        while (try stmt.step()) {
            try progress_list.append(ReadingProgress{
                .id = stmt.columnInt64(0),
                .user_id = stmt.columnInt64(1),
                .book_id = stmt.columnInt64(2),
                .current_page = stmt.columnInt64(3),
                .total_pages = if (stmt.columnType(4) == .Null) null else stmt.columnInt64(4),
                .percentage_complete = stmt.columnDouble(5),
                .last_read_at = stmt.columnInt64(6),
                .completed_at = if (stmt.columnType(7) == .Null) null else stmt.columnInt64(7),
                .created_at = stmt.columnInt64(8),
                .updated_at = stmt.columnInt64(9),
            });
        }

        return progress_list;
    }

    pub fn getCompletedBooks(self: *ReadingProgressService, user_id: i64) !std.ArrayList(ReadingProgress) {
        var progress_list = std.ArrayList(ReadingProgress).init(self.allocator);
        errdefer progress_list.deinit();

        const query =
            \\SELECT id, user_id, book_id, current_page, total_pages, percentage_complete, last_read_at, completed_at, created_at, updated_at
            \\FROM reading_progress
            \\WHERE user_id = ? AND completed_at IS NOT NULL
            \\ORDER BY completed_at DESC
        ;

        var stmt = try self.db.prepare(query);
        defer stmt.deinit();

        try stmt.bind(.{user_id});

        while (try stmt.step()) {
            try progress_list.append(ReadingProgress{
                .id = stmt.columnInt64(0),
                .user_id = stmt.columnInt64(1),
                .book_id = stmt.columnInt64(2),
                .current_page = stmt.columnInt64(3),
                .total_pages = if (stmt.columnType(4) == .Null) null else stmt.columnInt64(4),
                .percentage_complete = stmt.columnDouble(5),
                .last_read_at = stmt.columnInt64(6),
                .completed_at = if (stmt.columnType(7) == .Null) null else stmt.columnInt64(7),
                .created_at = stmt.columnInt64(8),
                .updated_at = stmt.columnInt64(9),
            });
        }

        return progress_list;
    }

    pub const ReadingStats = struct {
        total_books: i64,
        completed_books: i64,
        currently_reading: i64,
        total_pages_read: i64,
    };

    pub fn getReadingStats(self: *ReadingProgressService, user_id: i64) !ReadingStats {
        const query =
            \\SELECT 
            \\  COUNT(*) as total_books,
            \\  SUM(CASE WHEN completed_at IS NOT NULL THEN 1 ELSE 0 END) as completed_books,
            \\  SUM(CASE WHEN completed_at IS NULL AND current_page > 0 THEN 1 ELSE 0 END) as currently_reading,
            \\  SUM(current_page) as total_pages_read
            \\FROM reading_progress
            \\WHERE user_id = ?
        ;

        var stmt = try self.db.prepare(query);
        defer stmt.deinit();

        try stmt.bind(.{user_id});

        if (try stmt.step()) {
            return ReadingStats{
                .total_books = stmt.columnInt64(0),
                .completed_books = stmt.columnInt64(1),
                .currently_reading = stmt.columnInt64(2),
                .total_pages_read = stmt.columnInt64(3),
            };
        }

        return ReadingStats{
            .total_books = 0,
            .completed_books = 0,
            .currently_reading = 0,
            .total_pages_read = 0,
        };
    }
};
